dbt/adapters/bigquery.py

```diff
@@ -0,0 +1,300 @@
+
+import dbt.exceptions
+import dbt.flags as flags
+import dbt.materializers
+import dbt.clients.gcloud
+
+from dbt.adapters.postgres import PostgresAdapter
+from dbt.contracts.connection import validate_connection
+from dbt.logger import GLOBAL_LOGGER as logger
+
+
+class BigQueryAdapter(PostgresAdapter):
+
+    QUERY_TIMEOUT = 60 * 1000
+    requires = {'bigquery': 'google-cloud-bigquery==0.24.0'}
+
+    @classmethod
+    def initialize(cls):
+        google = cls._import('google')
+        google.auth = cls._import('google.auth')
+        google.oauth2 = cls._import('google.oauth2')
+
+        google.cloud = cls._import('google.cloud')
+        google.cloud.bigquery = cls._import('google.cloud.bigquery')
+        google.cloud.exceptions = cls._import('google.cloud.exceptions')
+
+        globals()['google'] = google
+
+    @classmethod
+    def get_materializer(cls, node, existing):
+        materializer = dbt.materializers.NonDDLMaterializer
+        return dbt.materializers.make_materializer(materializer,
+                                                   cls,
+                                                   node,
+                                                   existing)
+
+    @classmethod
+    def handle_error(cls, error, message, sql):
+        logger.debug(message.format(sql=sql))
+        logger.debug(error)
+        error_msg = "\n".join([error['message'] for error in error.errors])
+        raise dbt.exceptions.RuntimeException(error_msg)
+
+    @classmethod
+    @contextmanager
+    def exception_handler(cls, profile, sql, model_name=None,
+                          connection_name='master'):
+        try:
+            yield
+
+        except google.cloud.exceptions.BadRequest as e:
+            message = "Bad request while running:\n{sql}"
+            cls.handle_error(e, message, sql)
+
+        except google.cloud.exceptions.Forbidden as e:
+            message = "Access denied while running:\n{sql}"
+            cls.handle_error(e, message, sql)
+
+        except Exception as e:
+            logger.debug("Unhandled error while running:\n{}".format(sql))
+            logger.debug(e)
+            raise dbt.exceptions.RuntimeException(e)
+
+    @classmethod
+    def type(cls):
+        return 'bigquery'
+
+    @classmethod
+    def date_function(cls):
+        return 'CURRENT_TIMESTAMP()'
+
+    @classmethod
+    def begin(cls, profile, name='master'):
+        pass
+
+    @classmethod
+    def commit(cls, connection):
+        pass
+
+    @classmethod
+    def get_status(cls, cursor):
+        raise Exception("Not implemented")
+        state = cursor.sqlstate
+
+        if state is None:
+            state = 'SUCCESS'
+
+        return "{} {}".format(state, cursor.rowcount)
+
+    @classmethod
+    def get_bigquery_credentials(cls, config):
+        method = config.get('method')
+        Creds = google.oauth2.service_account.Credentials
+
+        if method == 'oauth':
+            return None
```

Comment:
We initialize a bigquery client with the `credentials` argument if it's a service account (from a private key file or the private key raw json).

If the credentials object is `None` (as with oauth), then the bq client lib will use the oauth flow to authenticate the user. We force this flow if the user isn't authenticated by shelling out to `gcloud` [here](https://github.com/fishtown-analytics/dbt/pull/455/files#diff-a8966822c3061ac7ef67f0f6416e5cd8R135)

More info on gcp auth flows [here](https://googlecloudplatform.github.io/google-cloud-python/stable/google-cloud-auth.html#credential-discovery-precedence)